core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java

```diff
@@ -207,68 +214,97 @@ private Object prefetchKeysIfNeededAndInvokeNext(InvocationContext ctx, Visitabl
                nil -> makeStage(prefetchKeysIfNeededAndInvokeNext(ctx, command, keys, true)).toCompletableFuture()));
       }
       if (transferedKeys != null) {
-         return asyncInvokeNext(ctx, command, retrieveRemoteValues(ctx, transferedKeys).toCompletableFuture());
+         return asyncInvokeNext(ctx, command, retrieveRemoteValues(ctx, transferedKeys, command.getTopologyId()).toCompletableFuture());
       } else {
          return invokeNext(ctx, command);
       }
    }
 
-   private InvocationStage retrieveSingleRemoteValue(InvocationContext ctx, Object key) {
+   private InvocationStage retrieveSingleRemoteValue(InvocationContext ctx, Object key, int topologyId) {
```

Comment:
Right, it sounds a bit confusing. `retrieveSingleRemoteValue` runs first the local lookup (for the case that we already have the key) and then invokes `retrieveRemoteValue`. How would you like to call this method which does first local and then remote lookup? Just `retrieveSingleValue`?

Writing this comment I've realized that it may not work properly, though; Using the RemoteMetadata is correct only during the value transfer, but during key transfer the metadata could point to a node with outdated copy. I'll change that.